src/ai/defaultai.cc

```diff
@@ -3342,7 +3342,7 @@ void DefaultAI::diplomacy_actions(const Time& gametime) {
 
 	const Widelands::PlayerNumber mypn = player_number();
 	const Widelands::Player* me = game().get_player(mypn);
-	const Widelands::TeamNumber mytn = me->team_number();
+	const Widelands::TeamNumber mytn = player_statistics.get_team_number(mypn);
```

Comment:
wouldn't this mean we are using outdated team information. As our team number might have been modified in the mean time by another player. Would it be good to base the evaluation forther down on such data.

Reply:
> wouldn't this mean we are using outdated team information. As our team number might have been modified in the mean time by another player. Would it be good to base the evaluation forther down on such data.

My reasoning was that if we can't prevent stale data, at least we should use it consistently. I thought I left a comment to that effect…
 
> Edit: Furthermore I believe team_number wasn't the problem in the case from the log. me_alone is the culprit here.

It's both of them interacting. The stale team number would have been 0, so leave wouldn't have been triggered.

> Edit2: I think the best solution for this issue is to recalculate team powers after a team change has been done. And to make this safe for the interaction with humans this might be best done in Diplomacy action code.

Actually the best solution would be to change the main diplomacy code to invalidate all pending requests by or to a player when their team changes. So again, out of scope for AI Diplo.

Reply:
> > wouldn't this mean we are using outdated team information. As our team number might have been modified in the mean time by another player. Would it be good to base the evaluation forther down on such data.
> 
> My reasoning was that if we can't prevent stale data, at least we should use it consistently. I thought I left a comment to that effect…    
   
understood, but I still beleive this would not help for the specific issue. 
      
> 
> > Edit2: I think the best solution for this issue is to recalculate team powers after a team change has been done. And to make this safe for the interaction with humans this might be best done in Diplomacy action code.
> 
> Actually the best solution would be to change the main diplomacy code to delete all pending requests by or to a player when their team changes. So again, out of scope for AI Diplo.         
    
well Leaving is not dependent on a open action isn't it.       
   
We should ask @Noordfrees whether we can call a function defined for a specific computer player (namely the player_statistics.recalculate_team_powers() function as defined in default_ai.h  (defining the PLayers_strenghts Struct for the computerplayer) and ai_help_structs.h (defining the actual function of the struct)) from the CmdDiplomacy::execute() code

Reply:
> understood, but I still beleive this would not help for the specific issue.

That's why I called it a mitigation, not a solution. :)
 
> well Leaving is not dependent on a open action isn't it.

True, but at least it's halfway detectable by comparing the current team with the stored one. Later on we have much less chance to detect outdated stats.
 
> We should ask @Noordfrees whether we can call a function defined for a specific computer player (namely the player_statistics.recalculate_team_powers() function as defined in default_ai.h (defining the PLayers_strenghts Struct for the computerplayer) and ai_help_structs.h (defining the actual function of the struct)) from the CmdDiplomacy::execute() code

I don't think such complications are needed. Only loop through pending actions right there in the accept handling code, and remove matching ones. Pending actions is always as fresh as it can be on our side.

Reply:
Well I understood the problem above as Player 1 invited Player 2 and afterwards missed the acception of Player 2. So he left the team shortly after. I can't see any open pending action involved here.  
Only thing that should help is to recalculate all values realted to teams as soon as we change the teams. And as a team change can happen due to a human player as well we need to find a way to trigger the recalculation after every team change. 
for the current problem it would have been suitable to recalculate after the player 2 accepted. so as a first mitigation we might at least recalculate after an Ai triggered team change.

Reply:
Outdated pending actions are a similar possibility. I'm almost sure I saw that too in some of my tests. It's also something that is a more general issue (papered over for AIs by this PR actually), that also affects human players, even more than AIs, because they may postpone dealing with diplomacy requests for much longer. So this is worth solving in general, by the main diplomacy code. (at least I believe it's going to be simple)

As you said, these race conditions in `diplomacy_actions()` are not all that important, so we shouldn't go out of our way to solve them, only apply simple mitigations where possible, or just ignore them. So yes, please merge this as it is now (without mitigation) and then we can discuss more in the main PR if needed.